apps/hannk/interpreter/transforms.cpp

```diff
@@ -27,67 +31,93 @@ T *cast_op(Op *x) {
     }
 }
 
-class RemoveDeadOps {
-    const Op *const root_;
+class RemoveDeadOps : public OpMutator {
+    using OpMutator::visit;
+
+    std::unordered_set<Tensor *> root_outputs_;
+    int removed_ = 0;
 
     bool is_root_output(const TensorPtr &t) const {
-        return root_->is_output(t);
+        return root_outputs_.count(t.get()) > 0;
     };
 
-public:
-    explicit RemoveDeadOps(Op *root)
-        : root_(root) {
-    }
-
-    void remove_in_group(OpGroup *op_group) {
-        // Find ops with outputs that are unused.
-        // Go in reverse order so removing a dead op
-        // enables earlier ops to be seen as dead.
-        for (int i = op_group->op_count() - 1; i >= 0; --i) {
-            Op *op = op_group->op(i);
-            if (OpGroup *group = cast_op<OpGroup>(op)) {
-                remove_in_group(group);
+    bool is_dead(const Op *op) const {
+        for (int j = 0; j < op->input_count(); j++) {
+            const auto &input = op->input(j);
+            if (is_root_output(input)) {
+                // TODO: is it ever actually possible to have an Op's input be a root output?
+                // (This is what the previous, OpVisitor-based code did)
+                return false;
             }
-            bool dead = true;
-            for (int j = 0; dead && j < op->input_count(); j++) {
-                if (is_root_output(op->input(j))) {
-                    dead = false;
-                    break;
-                }
+        }
+        for (int j = 0; j < op->output_count(); j++) {
+            const auto &output = op->output(j);
+            // An op isn't dead if its output is an output
+            // of the graph.
+            if (is_root_output(output)) {
+                return false;
             }
-            for (int j = 0; dead && j < op->output_count(); j++) {
-                // An op isn't dead if its output is an output
-                // of the graph.
-                if (is_root_output(op->output(j))) {
-                    dead = false;
-                    break;
-                }
 
-                if (!op->output(j)->consumers().empty()) {
-                    dead = false;
-                    break;
-                }
+            if (!output->consumers().empty()) {
+                return false;
             }
+        }
+        return true;
+    }
 
-            if (dead) {
-                op_group->remove(op);
-            }
+public:
+    // Go in reverse order so removing a dead op enables earlier ops to be seen as dead.
+    explicit RemoveDeadOps(const Op *root)
+        : OpMutator(OpMutator::Reverse) {
```

Comment:
Is the reverse order a common use case? I am wondering if RemoveDeadOps should just override the default behavior rather than building it into the base class.

Reply:
No, it has exactly one use case currently. I initially did what you suggest here, but the implementation of `OpMutator::visit` ended up feeling ticklish enough to get right that I thought it more prudent to encapsulate it all in one spot, rather than encourage possibly-wrong code replication. (I don't feel strongly about this, though.)

Reply:
PTAL